core/src/docs/rfcs/5444_operator_from_uri.md

```diff
@@ -0,0 +1,144 @@
+- Proposal Name: `operator_from_uri`
+- Start Date: 2024-12-23
+- RFC PR: [apache/opendal#5444](https://github.com/apache/opendal/pull/5444)
+- Tracking Issue: [apache/opendal#5445](https://github.com/apache/opendal/issues/5445)
+
+# Summary
+
+This RFC proposes adding URI-based configuration support to OpenDAL, allowing users to create operators directly from URIs. The proposal introduces a new `from_uri` API in both the `Operator` and `Configurator` traits, along with an `OperatorRegistry` to manage operator factories.
+
+# Motivation
+
+Currently, creating an operator in OpenDAL requires explicit configuration through builder patterns. While this approach provides type safety and clear documentation, it can be verbose and inflexible for simple use cases. Many storage systems are naturally identified by URIs (e.g., `s3://bucket/path`, `fs:///path/to/dir`).
+
+Adding URI-based configuration would:
+
+- Simplify operator creation for common use cases
+- Enable configuration via connection strings (common in many applications)
+- Make OpenDAL more approachable for new users
+- Allow dynamic operator creation based on runtime configuration
+
+# Guide-level explanation
+
+The new API allows creating operators directly from URIs:
+
+```rust
+// Create an operator using URI
+let op = Operator::from_uri("s3://my-bucket/path", vec![
+    ("access_key_id".to_string(), "xxx".to_string()),
+    ("secret_key_key".to_string(), "yyy".to_string()),
+])?;
+
+// Create a file system operator
+let op = Operator::from_uri("fs:///tmp/test", vec![])?;
+
+// Using with custom registry
+let registry = OperatorRegistry::new();
+registry.register("custom", my_factory);
+let op = registry.parse("custom://endpoint", options)?;
+```
+
+# Reference-level explanation
+
+The implementation consists of three main components:
+
+1. The `OperatorRegistry`:
+
+```rust
+type OperatorFactory = fn(http::Uri, HashMap<String, String>) -> Result<Operator>;
+
+pub struct OperatorRegistry {
+    register: Arc<Mutex<HashMap<String, OperatorFactory>>>,
+}
+
+impl OperatorRegistry {
+    fn register(&self, scheme: &str, factory: OperatorFactory) {
+        ...
+    }
+
+    fn parse(&self, uri: &str, options: impl IntoIterator<Item = (String, String)>) -> Result<Operator> {
+        ...
+    }
+}
+```
+
+2. The `Configurator` trait extension:
+
+```rust
+impl Configurator for S3Config {
+    fn from_uri(uri: &str, options: impl IntoIterator<Item = (String, String)>) -> Result<Self> {
+        ...
+    }
+}
+```
+
+3. The `Operator` factory method:
+
+```rust
+impl Operator {
+    pub fn from_uri(
+        uri: &str,
+        options: impl IntoIterator<Item = (String, String)>,
+    ) -> Result<Self> {
+        static REGISTRY: Lazy<OperatorRegistry> = Lazy::new(|| {
+            let registry = OperatorRegistry::new();
+            // Register built-in operators
+            registry.register("s3", s3_factory);
```

Comment:
where will those factory be defined? as static variables? 

I think we should declare a `S3Factory` unit struct and use

```suggestion
            registry.register("s3", S3Factory);
```

instead. I don't like to have that much static variables

Reply:
> where will those factory be defined?

They are functions defined by users.

> as static variables?

No.

> I think we should declare a `S3Factory` unit struct and use

It's not defined by `opendal`. I will improve the example to make it clearer.

Reply:
Oh, I thought `opendal` would expose an `OperatorRegistry` with all the scheme mappings.

Should every user define their OperatorRegistry then?

Reply:
> Should every user define their OperatorRegistry then?

We will have a gloabl one (a.k.a the `Operator::from_uri`). Users who want to register their own can create a `OperatorRegistry` instead.

Reply:
Awesome then! sorry for my misunderstood, I lack some context about the project. Thank you very much for your answers